weave/trace_server/calls_query_builder.py

```diff
@@ -0,0 +1,726 @@
+        res = ""
+        for index, (cast, direction) in enumerate(options):
+            if index > 0:
+                res += ", "
+            res += f"{self.field.as_sql(pb, table_alias, cast)} {direction}"
+        return res
+
+
+class Condition(BaseModel):
+    operand: "tsi_query.Operand"
+    _consumed_fields: typing.Optional[list[CallsMergedField]] = None
+
+    def as_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        conditions = process_query_to_conditions(
+            tsi_query.Query.model_validate({"$expr": {"$and": [self.operand]}}),
+            pb,
+            table_alias,
+        )
+        if self._consumed_fields is None:
+            self._consumed_fields = []
+            for field in conditions.fields_used:
+                self._consumed_fields.append(get_field_by_name(field))
+        return combine_conditions(conditions.conditions, "AND")
+
+    def _get_consumed_fields(self) -> list[CallsMergedField]:
+        if self._consumed_fields is None:
+            self.as_sql(ParamBuilder(), "calls_merged")
+        if self._consumed_fields is None:
+            raise ValueError("Consumed fields should not be None")
+        return self._consumed_fields
+
+    def is_heavy(self) -> bool:
+        for field in self._get_consumed_fields():
+            if field.is_heavy():
+                return True
+        return False
+
+
+class HardCodedFilter(BaseModel):
+    filter: tsi._CallsFilter
+
+    def is_useful(self) -> bool:
+        return any(
+            [
+                self.filter.op_names,
+                self.filter.input_refs,
+                self.filter.output_refs,
+                self.filter.parent_ids,
+                self.filter.trace_ids,
+                self.filter.call_ids,
+                self.filter.trace_roots_only,
+                self.filter.wb_user_ids,
+                self.filter.wb_run_ids,
+            ]
+        )
+
+    def as_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        return combine_conditions(
+            process_calls_filter_to_conditions(self.filter, pb, table_alias),
+            "AND",
+        )
+
+
+class CallsQuery(BaseModel):
+    """Critical to be injection safe!"""
+
+    project_id: str
+    select_fields: list[CallsMergedField] = Field(default_factory=list)
+    query_conditions: list[Condition] = Field(default_factory=list)
+    hardcoded_filter: typing.Optional[HardCodedFilter] = None
+    order_fields: list[OrderField] = Field(default_factory=list)
+    limit: typing.Optional[int] = None
+    offset: typing.Optional[int] = None
+
+    def add_field(self, field: str) -> "CallsQuery":
+        self.select_fields.append(get_field_by_name(field))
+        return self
+
+    def add_condition(self, operand: "tsi_query.Operand") -> "CallsQuery":
+        if isinstance(operand, tsi_query.AndOperation):
+            if len(operand.and_) == 0:
+                raise ValueError("Empty AND operation")
+            else:
```

Comment:
Stylistically I prefer omitting the else clause when you have a raise like this, but up to you.

Reply:
fixed